blockchain/contracts/Staking.sol

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+import "@openzeppelin/contracts/interfaces/IERC721.sol";
+import "@openzeppelin/contracts/interfaces/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract Staking is IERC721Receiver {
+    // TODO natspec
+    // TODO emergencyWithdrawal, allocateRewards
+    IERC721 public characterToken;
+    IERC20 public gemsToken;
+    uint256 public STAKING_FEE = 100 * 10**18;
+
+    struct Stake {
+        uint256 tokenId;
+        uint256 gemsAmount;
+        uint256 timestamp;
+    }
+
+    event NFTStaked(address owner, uint256 tokenId, uint256 gemsAmount);
+    event NFTUnstaked(address owner, uint256 tokenId, uint256 rewardedAmount);
+    event Claimed(address owner);
+
+    // Map staker address to stake details
+    mapping(address => Stake) public stakes;
+    // Easier to manage than verifying the content of the Stake Struct which can be tricky due to deletion
+    mapping(address => bool) public isStaking;
+
+    // Allows only one NFT staked at a time per user
+    modifier onlyOneStake() {
+        require(!isStaking[msg.sender], "Only one stake is allowed at once");
+        _;
+    }
+
+    modifier minimumAmount(uint256 gemsAmount) {
+        require(
+            gemsAmount >= STAKING_FEE && gemsBalance(msg.sender) >= gemsAmount,
+            "You need to have at least 100 gems to stake"
```

Comment:
nit: 'Insufficient amount of gems. You need to have at least 100 gems to stake.'

Reply:
Agreed, i'll add the Insufficient amount: